### PR TITLE
IAM takes a while to propagate new keys

### LINKED ISSUE #132: Key rotation: documentation and/or bug
I'm attempting to `aws-vault rotate <profile>` where profile is an IAM user with AdministratorAccess policy attached, and I'm seeing something like https://github.com/99designs/aws-vault/issues/15:

```
$ aws-vault rotate pda
Enter passphrase to unlock /home/pda/.awsvault/keys:
Using old credentials to create a new access key
aws-vault: error: InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: ...
```

Using a `v3.7.1-23-g8c008d3` build of aws-vault.

Same error for any `aws iam ...` operation, but all other AWS services are working;

```
$ aws-vault exec pda -- aws iam list-access-keys
Enter passphrase to unlock /home/pda/.awsvault/keys:
An error occurred (InvalidClientTokenId) when calling the ListAccessKeys operation: The security token included in the request is invalid
```

Is this expected? Is there an IAM setting/policy somewhere that needs to be changed? Or does rotation only work for non-IAM root account credentials? Should we be using the stored access key rather than a session token for the `rotate` command?

I've used `aws-vault rotate` before on work AWS accounts, but never with this personal AWS account.

As https://github.com/99designs/aws-vault/issues/129 pointed out, there's no mention of `aws-vault rotate` in the README etc; I'm happy to add that if I can figure out how to use it :)

### PR DESCRIPTION
**Pull request by lox**
This is why key rotation was failing, new IAM keys take a while to propagate. This adds a terrible sleep and then poll. 

Fixes #132.

### CODE REVIEW

File: cli/rotate.go [hunk 1]
```diff
@@ -142,20 +150,26 @@ func RotateCommand(app *kingpin.Application, input RotateCommandInput) {
 		return
 	}
 
-	newVal, err := newSessionCreds.Get()
-	if err != nil {
-		app.Fatalf(awsConfig.FormatCredentialError(err, input.Profile))
-		return
-	}
+	log.Printf("Waiting for new IAM credentials to propagate")
+	time.Sleep(time.Second * 8)
```
Comment (mtibben) [inline cli/rotate.go]: 8 seconds seems a while, should we just start polling immediately?

File: cli/login.go [hunk 2]
```diff
@@ -219,11 +220,20 @@ func getFederationToken(creds credentials.Value, d time.Duration) (*sts.Credenti
...
+
 func getCurrentUserName(sess *session.Session) (string, error) {
 	client := iam.New(sess)
 
 	resp, err := client.GetUser(&iam.GetUserInput{})
 	if err != nil {
+		matches := getUserErrorRegexp.FindAllStringSubmatch(err.Error(), -1)
+		if len(matches) > 0 {
+			log.Printf("%#v", matches)
+		}
```
Comment (mtibben) [inline cli/login.go]: What is this logging? does it need a prefix?

File: cli/rotate.go (hunk 1 above)
Comment (lox) [reply cli/rotate.go]: It seems pretty consistently to take about 8 seconds :)

Comment (lox) [reply cli/rotate.go]: We could suppress errors and poll regularly though.

File: cli/login.go (hunk 2 above)
Comment (lox) [reply cli/login.go]: Good catch, was purely debugging. Feels on pattern matching this error? It seems epic-ly el grosso.

File: cli/add.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ func ConfigureAddCommand(app *kingpin.Application) {
 func AddCommand(app *kingpin.Application, input AddCommandInput) {
 	var accessKeyId, secretKey string
 
-	if source, ok := awsConfig.SourceProfile(input.Profile); ok {
+	if source, _ := awsConfig.SourceProfile(input.Profile); source.Name != input.Profile {
```
Comment (lox) [inline cli/add.go]: This change is unrelated, and is just a bug I introduced in master.

File: cli/rotate.go [hunk 4]
```diff
@@ -92,13 +94,19 @@ func RotateCommand(app *kingpin.Application, input RotateCommandInput) {
 		return
 	}
 
+	var iamUserName *string
+
+	// A username is needed for some IAM calls if the credentials have assumed a role
```
Comment (lox) [inline cli/rotate.go]: Context here @mtibben is that for the root user this doesn't exist and needs to be null.

**Comment by lox** [comment]
I've moved everything to `vault.Rotator` @mtibben and cleaned things up a bit.

**Comment by mtibben** [comment]
lgtm

**Comment by damonmaria** [comment]
Thanks for the fix 👍 Been missing my key rotation :)

Would love an update in homebrew when someone with privileges gets the chance.

**Pull request merged**